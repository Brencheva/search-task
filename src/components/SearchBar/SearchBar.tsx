import { useGetProducts } from "../../hooks";
import ProductCard from "../ProductCard/ProductCard";
import { CloseIcon, SearchIcon } from "./SearchBar.const";

export default function SearchBar() {
  const { query, setQuery, loading, error, products } = useGetProducts();

  const search = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setQuery(value);
  };

  const reset = () => {
    setQuery("");
  };

  if (error) return <p>Error: {error}</p>;

  return (
    <div className="p-8 flex flex-col align-center">
      <div className="w-full flex items-center relative mb-8">
        <input
          type="text"
          placeholder="search"
          data-testid="search-field"
          className="border border-purple-400 rounded-lg py-4 px-12 w-full"
          onChange={search}
          value={query}
        />
        <SearchIcon />
        <button className="absolute right-0 mr-2 " onClick={reset} data-testid="reset-button">
          <CloseIcon />
        </button>
      </div>
      <ul
        className={
          products.length ? "list-none flex flex-wrap width-full" : "invisible"
        }
      >
        {products.map((product) => {
          return <ProductCard key={product.id} {...product} />;
        })}
      </ul>
      {!loading && !products.length && <p>No items found</p>}
      {loading && <p>Loading..</p>}
    </div>
  );
}
