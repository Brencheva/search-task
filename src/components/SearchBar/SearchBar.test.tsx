/* eslint-disable testing-library/prefer-screen-queries */
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent } from "@testing-library/react";
import SearchBar from "./SearchBar";

const query = { target: { value: "iphone" } };

describe("Search Bar", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<SearchBar />);
    expect(asFragment).toMatchSnapshot();
  });

  it("search field should have value", () => {
    const { getByTestId } = render(<SearchBar />);
    const searchField = getByTestId("search-field");
    fireEvent.change(searchField, query);
    expect(searchField).toHaveValue("iphone");
  });

  it("search reset button should reset the field", () => {
    const { getByTestId } = render(<SearchBar />);
    const searchField = getByTestId("search-field");
    const button = getByTestId("reset-button");
    fireEvent.change(searchField, query);
    expect(searchField).toHaveValue("iphone");
    button.click();

    setTimeout(() => {
      expect(searchField).toHaveValue("");
    }, 500);
  });
});
