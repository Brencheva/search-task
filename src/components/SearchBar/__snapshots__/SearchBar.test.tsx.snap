// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar renders correctly 1`] = `[Function]`;
